src/core_init_atmosphere/mpas_kd_tree.F

```diff
@@ -0,0 +1,394 @@
+
+
+   !***********************************************************************
+   !
+   !  recursive routine mpas_kd_search_internal
+   !
+   !> \brief   Recursively search the KD-Tree for query
+   !> \author  Miles A. Curry
+   !> \date    01/28/20
+   !> \details
+   !> Private, recursive function to search kdtree for query. Upon succes
+   !> res will point to the nearest neighbor to query and distance will hold
+   !> the squared distance between query and res.
+   !>
+   !> Distance is calculated and compared as squared distance to increase
+   !> efficiency.
+   !
+   !-----------------------------------------------------------------------
+   recursive subroutine mpas_kd_search_internal(kdtree, query, res, distance)
+
+      implicit none
+
+      ! Input Variables
+      type(mpas_kd_type), pointer, intent(in) :: kdtree
+      real(kind=RKIND), dimension(:), intent(in) :: query
+      type(mpas_kd_type), pointer, intent(inout) :: res
+      real(kind=RKIND), intent(inout) :: distance
+
+      ! Local Values
+      real(kind=RKIND) :: current_distance
+
+      current_distance = sum((kdtree % point(:) - query(:))**2)
+      if (current_distance < distance) then
+         distance = current_distance
+         res => kdtree
+      endif
+
+      !
+      ! To find the nearest neighbor, first serach the tree in a similar manner
+      ! as a single dimensioned BST, by comparing points on the current split
+      ! dimension.
+      !
+      ! If the distance between the current node and the query is less then the
+      ! minimum distance found within the subtree we just searched, then the nearest
+      ! neighbor might be in the opposite subtree, so search it.
+      !
+
+      if (query(kdtree % split_dim) > kdtree % point(kdtree % split_dim)) then
+         if (associated(kdtree % right)) then ! Search right
+            call mpas_kd_search_internal(kdtree % right, query, res, distance)
+         endif
+         if ((kdtree % point(kdtree % split_dim) - query(kdtree % split_dim))**2 <= distance .AND. associated(kdtree % left)) then
+            call mpas_kd_search_internal(kdtree % left, query, res, distance) ! Check the other subtree
+         endif
+      else if (query(kdtree % split_dim) < kdtree % point(kdtree % split_dim)) then
+         if (associated(kdtree % left)) then ! Search left
+            call mpas_kd_search_internal(kdtree % left, query, res, distance)
+         endif
+         if ((kdtree % point(kdtree % split_dim) - query(kdtree % split_dim))**2 <= distance .AND. associated(kdtree % right)) then
+            call mpas_kd_search_internal(kdtree % right, query, res, distance) ! Check the other subtree
+         endif
+      else ! Nearest point could be in either left or right subtree, so search both
+         if(associated(kdtree % right)) call mpas_kd_search_internal(kdtree % right, query, res, distance)
+         if(associated(kdtree % left)) call mpas_kd_search_internal(kdtree % left, query, res, distance)
```

Comment:
For consistency, I'd suggest not using a one-line if-test here, and instead using
```fortran
if (associated(kdtree % right)) then
   call mpas_kd_search_internal(kdtree % right, query, res, distance)
end if
```